File: mobile-apps/android/APIConfiguration.kt
/**
 * Fleet Management - Android API Configuration
 *
 * Unified API configuration for all mobile app features
 */

package com.capitaltechalliance.fleet

import android.content.Context
import android.os.Build
import java.util.UUID

enum class Environment {
    DEVELOPMENT,
    STAGING,
    PRODUCTION;

    companion object {
        val current: Environment
            get() = if (BuildConfig.DEBUG) DEVELOPMENT else PRODUCTION
    }
}

object APIConfiguration {
    // MARK: - Base URLs

    val baseURL: String
        get() = when (Environment.current) {
            Environment.DEVELOPMENT -> "http://10.0.2.2:3000" // Android emulator localhost
            Environment.STAGING -> "https://staging-fleet.capitaltechalliance.com"
            Environment.PRODUCTION -> "https://fleet.capitaltechalliance.com"
        }

    val apiURL: String
        get() = "$baseURL/api"

    val wsURL: String
        get() = when (Environment.current) {
            Environment.DEVELOPMENT -> "ws://10.0.2.2:3000"
            Environment.STAGING -> "wss://staging-fleet.capitaltechalliance.com"
            Environment.PRODUCTION -> "wss://fleet.capitaltechalliance.com"
        }

    // MARK: - Mobile Integration Endpoints

    object Mobile {
        const val register = "$apiURL/mobile/register"
        const val sync = "$apiURL/mobile/sync"
        fun route(vehicleId: Int) = "$apiURL/mobile/route/$vehicleId"
        const val arNavigation = "$apiURL/mobile/ar-navigation"
        const val keylessEntry = "$apiURL/mobile/keyless-entry"
        const val damageDetection = "$apiURL/mobile/damage-detection"
        const val dispatchMessages = "$apiURL/mobile/dispatch/messages"
        const val chargingStationsNearby = "$apiURL/mobile/charging-stations/nearby"
    }

    // MARK: - Telematics Endpoints

    object Telematics {
        const val providers = "$apiURL/telematics/providers"
        const val dashboard = "$apiURL/telematics/dashboard"
        fun vehicleLocation(vehicleId: Int) = "$apiURL/telematics/vehicles/$vehicleId/location"
        fun vehicleStats(vehicleId: Int) = "$apiURL/telematics/vehicles/$vehicleId/stats"
        const val safetyEvents = "$apiURL/telematics/safety-events"
    }

    // MARK: - Smartcar Endpoints

    object Smartcar {
        const val connect = "$apiURL/smartcar/connect"
        fun location(vehicleId: Int) = "$apiURL/smartcar/vehicles/$vehicleId/location"
        fun battery(vehicleId: Int) = "$apiURL/smartcar/vehicles/$vehicleId/battery"
        fun lock(vehicleId: Int) = "$apiURL/smartcar/vehicles/$vehicleId/lock"
        fun unlock(vehicleId: Int) = "$apiURL/smartcar/vehicles/$vehicleId/unlock"
        fun startCharging(vehicleId: Int) = "$apiURL/smartcar/vehicles/$vehicleId/charge/start"
        fun stopCharging(vehicleId: Int) = "$apiURL/smartcar/vehicles/$vehicleId/charge/stop"
    }

    // MARK: - Route Optimization Endpoints

    object Routes {
        const val optimize = "$apiURL/route-optimization/optimize"
        const val jobs = "$apiURL/route-optimization/jobs"
        const val active = "$apiURL/route-optimization/routes/active"
        fun route(routeId: Int) = "$apiURL/route-optimization/routes/$routeId"
        fun completeStop(routeId: Int, stopId: Int) = "$apiURL/route-optimization/routes/$routeId/stops/$stopId/complete"
    }

    // MARK: - Dispatch Endpoints

    object Dispatch {
        const val channels = "$apiURL/dispatch/channels"
        const val websocket = "$wsURL/api/dispatch/ws"
        fun history(channelId: Int) = "$apiURL/dispatch/channels/$channelId/history"
    }

    // MARK: - EV Charging Endpoints

    object EVCharging {
        const val chargers = "$apiURL/ev/chargers"
        fun chargerStatus(chargerId: Int) = "$apiURL/ev/chargers/$chargerId/status"
        fun reserve(chargerId: Int) = "$apiURL/ev/chargers/$chargerId/reserve"
        fun remoteStart(chargerId: Int) = "$apiURL/ev/chargers/$chargerId/remote-start"
        fun chargeSchedule(vehicleId: Int) = "$apiURL/ev/vehicles/$vehicleId/charge-schedule"
        const val carbonFootprint = "$apiURL/ev/carbon-footprint"
        const val esgReport = "$apiURL/ev/esg-report"
    }

    // MARK: - Video Telematics Endpoints

    object VideoTelematics {
        const val cameras = "$apiURL/video/cameras"
        const val events = "$apiURL/video/events"
        fun videoClip(eventId: Int) = "$apiURL/video/events/$eventId/clip"
        const val evidenceLocker = "$apiURL/video/evidence-locker"
    }

    // MARK: - 3D Vehicle Viewer Endpoints

    object Vehicle3D {
        fun model(vehicleId: Int) = "$apiURL/vehicles/$vehicleId/3d-model"
        fun arModel(vehicleId: Int) = "$apiURL/vehicles/$vehicleId/ar-model"
        fun customize(vehicleId: Int) = "$apiURL/vehicles/$vehicleId/customize"
        const val catalog = "$apiURL/vehicle-models/catalog"
    }

    // MARK: - Request Headers

    fun headers(token: String? = null, context: Context): Map<String, String> {
        val headers = mutableMapOf(
            "Content-Type" to "application/json",
            "Accept" to "application/json",
            "X-App-Version" to getAppVersion(context),
            "X-Device-Type" to "android",
            "X-OS-Version" to Build.VERSION.RELEASE
        )

        token?.let {
            headers["Authorization"] = "Bearer $it"
        }

        return headers
    }

    // MARK: - Device Info

    fun getAppVersion(context: Context): String {
        return try {
            val packageInfo = context.packageManager.getPackageInfo(context.packageName, 0)
            packageInfo.versionName ?: "1.0.0"
        } catch (e: Exception) {
            "1.0.0"
        }
    }

    fun getDeviceId(context: Context): String {
        val sharedPrefs = context.getSharedPreferences("fleet_prefs", Context.MODE_PRIVATE)
        var deviceId = sharedPrefs.getString("device_id", null)

        if (deviceId == null) {
            deviceId = UUID.randomUUID().toString()
            sharedPrefs.edit().putString("device_id", deviceId).apply()
        }

        return deviceId
    }

    fun getDeviceName(): String {
        return "${Build.MANUFACTURER} ${Build.MODEL}"
    }

    // MARK: - Timeout Configuration

    const val REQUEST_TIMEOUT = 30_000L // 30 seconds
    const val UPLOAD_TIMEOUT = 120_000L // 2 minutes
    const val DOWNLOAD_TIMEOUT = 120_000L // 2 minutes

    // MARK: - Feature Flags

    object FeatureFlags {
        const val offlineMode = true
        const val barcodeScanner = true
        const val aiDamageDetection = true
        const val lidarScanning = false // LiDAR not available on Android
        const val keylessEntry = true
        const val arNavigation = true
        const val radioDispatch = true
        const val videoTelematics = true
        const val evCharging = true
        const val routeOptimization = true
    }

    // MARK: - Offline Sync Configuration

    object OfflineSync {
        const val AUTO_SYNC_INTERVAL_MINUTES = 15
        const val MAX_RETRY_ATTEMPTS = 5
        const val RETRY_BACKOFF_MULTIPLIER = 2.0
        const val MAX_CONCURRENT_SYNC_OPERATIONS = 3
    }

    // MARK: - Cache Configuration

    object Cache {
        const val VEHICLES_CACHE_DURATION_MS = 60 * 60 * 1000L // 1 hour
        const val ROUTES_CACHE_DURATION_MS = 5 * 60 * 1000L // 5 minutes
        const val CHARGERS_CACHE_DURATION_MS = 15 * 60 * 1000L // 15 minutes
        const val MAX_PHOTO_CACHE_SIZE_MB = 100 // 100 MB
    }
}

// MARK: - OkHttp Client Extension

import okhttp3.OkHttpClient
import java.util.concurrent.TimeUnit

object FleetHttpClient {
    val client: OkHttpClient by lazy {
        OkHttpClient.Builder()
            .connectTimeout(APIConfiguration.REQUEST_TIMEOUT, TimeUnit.MILLISECONDS)
            .readTimeout(APIConfiguration.DOWNLOAD_TIMEOUT, TimeUnit.MILLISECONDS)
            .writeTimeout(APIConfiguration.UPLOAD_TIMEOUT, TimeUnit.MILLISECONDS)
            .retryOnConnectionFailure(true)
            .build()
    }
}

File: mobile-apps/ios/APIConfiguration.swift
/**
 * Fleet Management - iOS API Configuration
 *
 * Unified API configuration for all mobile app features:
 * - Offline storage and sync
 * - Driver toolbox
 * - Keyless entry
 * - AR navigation
 * - Barcode scanning
 * - AI damage detection
 * - LiDAR 3D scanning
 * - Route optimization
 * - Radio dispatch
 * - Video telematics
 * - EV charging
 */

import Foundation

enum Environment {
    case development
    case staging
    case production

    static var current: Environment {
        #if DEBUG
        return .development
        #else
        return .production
        #endif
    }
}

struct APIConfiguration {
    // MARK: - Base URLs

    static var baseURL: String {
        switch Environment.current {
        case .development:
            return "http://localhost:3000"
        case .staging:
            return "https://staging-fleet.capitaltechalliance.com"
        case .production:
            return "https://fleet.capitaltechalliance.com"
        }
    }

    static var apiURL: String {
        return "\(baseURL)/api"
    }

    static var wsURL: String {
        switch Environment.current {
        case .development:
            return "ws://localhost:3000"
        case .staging:
            return "wss://staging-fleet.capitaltechalliance.com"
        case .production:
            return "wss://fleet.capitaltechalliance.com"
        }
    }

    // MARK: - Mobile Integration Endpoints

    struct Mobile {
        static let register = "\(apiURL)/mobile/register"
        static let sync = "\(apiURL)/mobile/sync"
        static func route(vehicleId: Int) -> String {
            "\(apiURL)/mobile/route/\(vehicleId)"
        }
        static let arNavigation = "\(apiURL)/mobile/ar-navigation"
        static let keylessEntry = "\(apiURL)/mobile/keyless-entry"
        static let damageDetection = "\(apiURL)/mobile/damage-detection"
        static let dispatchMessages = "\(apiURL)/mobile/dispatch/messages"
        static let chargingStationsNearby = "\(apiURL)/mobile/charging-stations/nearby"
    }

    // MARK: - Telematics Endpoints

    struct Telematics {
        static let providers = "\(apiURL)/telematics/providers"
        static let dashboard = "\(apiURL)/telematics/dashboard"
        static func vehicleLocation(vehicleId: Int) -> String {
            "\(apiURL)/telematics/vehicles/\(vehicleId)/location"
        }
        static func vehicleStats(vehicleId: Int) -> String {
            "\(apiURL)/telematics/vehicles/\(vehicleId)/stats"
        }
        static let safetyEvents = "\(apiURL)/telematics/safety-events"
    }

    // MARK: - Smartcar Endpoints

    struct Smartcar {
        static let connect = "\(apiURL)/smartcar/connect"
        static func location(vehicleId: Int) -> String {
            "\(apiURL)/smartcar/vehicles/\(vehicleId)/location"
        }
        static func battery(vehicleId: Int) -> String {
            "\(apiURL)/smartcar/vehicles/\(vehicleId)/battery"
        }
        static func lock(vehicleId: Int) -> String {
            "\(apiURL)/smartcar/vehicles/\(vehicleId)/lock"
        }
        static func unlock(vehicleId: Int) -> String {
            "\(apiURL)/smartcar/vehicles/\(vehicleId)/unlock"
        }
        static func startCharging(vehicleId: Int) -> String {
            "\(apiURL)/smartcar/vehicles/\(vehicleId)/charge/start"
        }
        static func stopCharging(vehicleId: Int) -> String {
            "\(apiURL)/smartcar/vehicles/\(vehicleId)/charge/stop"
        }
    }

    // MARK: - Route Optimization Endpoints

    struct Routes {
        static let optimize = "\(apiURL)/route-optimization/optimize"
        static let jobs = "\(apiURL)/route-optimization/jobs"
        static let active = "\(apiURL)/route-optimization/routes/active"
        static func route(routeId: Int) -> String {
            "\(apiURL)/route-optimization/routes/\(routeId)"
        }
        static func completeStop(routeId: Int, stopId: Int) -> String {
            "\(apiURL)/route-optimization/routes/\(routeId)/stops/\(stopId)/complete"
        }
    }

    // MARK: - Dispatch Endpoints

    struct Dispatch {
        static let channels = "\(apiURL)/dispatch/channels"
        static let websocket = "\(wsURL)/api/dispatch/ws"
        static func history(channelId: Int) -> String {
            "\(apiURL)/dispatch/channels/\(channelId)/history"
        }
    }

    // MARK: - EV Charging Endpoints

    struct EVCharging {
        static let chargers = "\(apiURL)/ev/chargers"
        static func chargerStatus(chargerId: Int) -> String {
            "\(apiURL)/ev/chargers/\(chargerId)/status"
        }
        static func reserve(chargerId: Int) -> String {
            "\(apiURL)/ev/chargers/\(chargerId)/reserve"
        }
        static func remoteStart(chargerId: Int) -> String {
            "\(apiURL)/ev/chargers/\(chargerId)/remote-start"
        }
        static func chargeSchedule(vehicleId: Int) -> String {
            "\(apiURL)/ev/vehicles/\(vehicleId)/charge-schedule"
        }
        static let carbonFootprint = "\(apiURL)/ev/carbon-footprint"
        static let esgReport = "\(apiURL)/ev/esg-report"
    }

    // MARK: - Video Telematics Endpoints

    struct VideoTelematics {
        static let cameras = "\(apiURL)/video/cameras"
        static let events = "\(apiURL)/video/events"
        static func videoClip(eventId: Int) -> String {
            "\(apiURL)/video/events/\(eventId)/clip"
        }
        static let evidenceLocker = "\(apiURL)/video/evidence-locker"
    }

    // MARK: - 3D Vehicle Viewer Endpoints

    struct Vehicle3D {
        static func model(vehicleId: Int) -> String {
            "\(apiURL)/vehicles/\(vehicleId)/3d-model"
        }
        static func arModel(vehicleId: Int) -> String {
            "\(apiURL)/vehicles/\(vehicleId)/ar-model"
        }
        static func customize(vehicleId: Int) -> String {
            "\(apiURL)/vehicles/\(vehicleId)/customize"
        }
        static let catalog = "\(apiURL)/vehicle-models/catalog"
    }

    // MARK: - Request Headers

    static func headers(withAuth token: String? = nil) -> [String: String] {
        var headers = [
            "Content-Type": "application/json",
            "Accept": "application/json",
            "X-App-Version": appVersion,
            "X-Device-Type": "ios",
            "X-OS-Version": osVersion
        ]

        if let token = token {
            headers["Authorization"] = "Bearer \(token)"
        }

        return headers
    }

    // MARK: - Device Info

    static var appVersion: String {
        Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "1.0.0"
    }

    static var osVersion: String {
        UIDevice.current.systemVersion
    }

    static var deviceId: String {
        UIDevice.current.identifierForVendor?.uuidString ?? UUID().uuidString
    }

    static var deviceName: String {
        UIDevice.current.name
    }

    // MARK: - Timeout Configuration

    static let requestTimeout: TimeInterval = 30.0
    static let uploadTimeout: TimeInterval = 120.0
    static let downloadTimeout: TimeInterval = 120.0

    // MARK: - Feature Flags

    struct FeatureFlags {
        static let offlineMode = true
        static let barcode Scanner = true
        static let aiDamageDetection = true
        static let lidarScanning = true
        static let keylessEntry = true
        static let arNavigation = true
        static let radioDispatch = true
        static let videoTelematics = true
        static let evCharging = true
        static let routeOptimization = true
    }

    // MARK: - Offline Sync Configuration

    struct OfflineSync {
        static let autoSyncInterval: TimeInterval = 15 * 60 // 15 minutes
        static let maxRetryAttempts = 5
        static let retryBackoffMultiplier = 2.0
        static let maxConcurrentSyncOperations = 3
    }

    // MARK: - Cache Configuration

    struct Cache {
        static let vehiclesCacheDuration: TimeInterval = 60 * 60 // 1 hour
        static let routesCacheDuration: TimeInterval = 5 * 60 // 5 minutes
        static let chargersCacheDuration: TimeInterval = 15 * 60 // 15 minutes
        static let maxPhotoCacheSize: Int = 100 * 1024 * 1024 // 100 MB
    }
}

// MARK: - API Client Extension

extension URLSession {
    static var fleet: URLSession {
        let configuration = URLSessionConfiguration.default
        configuration.timeoutIntervalForRequest = APIConfiguration.requestTimeout
        configuration.timeoutIntervalForResource = APIConfiguration.downloadTimeout
        configuration.requestCachePolicy = .reloadIgnoringLocalCacheData
        return URLSession(configuration: configuration)
    }
}
